Re-sync rgb()/rgba() implementation with the latest spec

weinig opened the pull request:
#### c61c3004d66ab6e219241918cfa49ce12427ea4a
<pre>
Re-sync rgb()/rgba() implementation with the latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=272942">https://bugs.webkit.org/show_bug.cgi?id=272942</a>

Reviewed by Darin Adler.

The CSS Color specification has changed since rgb()/rgba() was last
worked on. This updates things to support:

- Modified parsing rules for the modern syntax which now allow mixed &lt;number&gt;
  and &lt;percentage&gt; values.
- Changes to the types used by the r/g/b/alpha symbols used in the relative
  color form to always be CSS_NUMBER.
- Requirement that the relative form serialize as `color(srgb )`.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-rgb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    Update another good chunk of tests to now pass.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponent):
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionAlpha):
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponents):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersLegacyRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersModernRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw):
    Work to separate parsing and normalization, allowing much of normalization
    to be reused for all the parsing paths.

(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
    Remove distinction in parsing behavior between rgb() and rgba().

Canonical link: <a href="https://commits.webkit.org/277842@main">https://commits.webkit.org/277842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3e4f00cd2b4502752c67cc31c49e2f862c44500

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48730 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51691 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51035 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33878 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25471 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25588 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20948 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23062 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6786 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45002 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53326 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23777 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47142 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42246 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46076 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10729 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25847 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/24764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -346,161 +378,163 @@ static Color parseRelativeRGBParametersRaw(CSSParserTokenRange& args, ColorParse
+                return { };
+
+            if (!args.atEnd())
+                return { };
```
Maybe easier to see what’s happening here if we omit the blank lines.

Commit:
Re-sync rgb()/rgba() implementation with the latest spec
https://bugs.webkit.org/show_bug.cgi?id=272942

Reviewed by Darin Adler.

The CSS Color specification has changed since rgb()/rgba() was last
worked on. This updates things to support:

- Modified parsing rules for the modern syntax which now allow mixed <number>
  and <percentage> values.
- Changes to the types used by the r/g/b/alpha symbols used in the relative
  color form to always be CSS_NUMBER.
- Requirement that the relative form serialize as `color(srgb )`.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-rgb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    Update another good chunk of tests to now pass.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponent):
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionAlpha):
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponents):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersLegacyRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersModernRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw):
    Work to separate parsing and normalization, allowing much of normalization
    to be reused for all the parsing paths.

(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
    Remove distinction in parsing behavior between rgb() and rgba().

Canonical link: https://commits.webkit.org/277842@main
